Bugfix/fixing incomplete job object

rob-c opened the pull request:
A potential fix for #495 . This performs a much better/safer check against whether a job has subjobs associated with it and contains some minor refactoring of the SubJobXMLList class to keep subjob logic contained within there.

I'm keen to see how this runs on Jenkins, this should work but even still...

I'm keen to add a test for this before merging but would like to get this into 6.1.21 as it saves me some headaches when switching between branches and developing plugins.

Fixes #495
Edit:
Fixes #461

milliams commented on python/Ganga/Core/GangaRepository/GangaRepositoryXML.py:
```diff
@@ -225,7 +225,7 @@ def shutdown(self):
         logger = getLogger()
         logger.debug("Shutting Down GangaRepositoryLocal: %s" % self.registry.name)
-        for k in self._fully_loaded.keys():
+        for k in [_k for _k in self._fully_loaded.keys() if _k not in self.incomplete_objects]:
```
So there are objects in `self._fully_loaded` which are also in `self.incomplete_objects`? That seems bad.

rob-c replied:
I don't know it's a belt and braces check for now anyway, this may get relaxed.

this is just a short PR to fix the problems I'm seeing and see if they make worse ones (which given there's almost certainly a missing import/typo, these changes do)

milliams replied:
A belt and braces check here will just hide potentially large problems. If we think there could be a problem with the consistency of these two lists then we need to tackle that problem, not ignore it.

rob-c replied:
Yes, but we should never be risking user data on disk ever. I'm still paranoid that something will come along and ruin my jobs as it is still doing.

milliams commented on python/Ganga/Core/GangaRepository/GangaRepositoryXML.py:
```diff
@@ -352,11 +354,12 @@ def _read_master_cache(self):
-                    self._cached_cat[this_id] = this_cache[2]
-                    self._cached_cls[this_id] = this_cache[3]
-                    self._cached_obj[this_id] = this_cache[4]
+                    if this_cache[4] >= 0:
```
What is this protecting against? Is the `id` sometimes negative?

rob-c replied:
That is the time, and yes it is potentially -ve if something went wrong in a previous version of ganga were bad caches were still saved.

milliams commented on python/Ganga/Core/GangaRepository/SubJobXMLList.py:
```diff
@@ -517,3 +495,45 @@ def _private_display(self, reg_slice, this_format, default_width, markup):
         return ds
+
+    @staticmethod
+    def JobHasChildrenTest(jobDirectory):
```
Function name should start with lower case.

milliams commented on python/Ganga/Core/GangaRepository/SubJobXMLList.py:
```diff
@@ -517,3 +495,45 @@ def _private_display(self, reg_slice, this_format, default_width, markup):
+        if count:
+            return True
+        else:
+            return False
```
This could just be:
```python
@staticmethod
def jobHasChildrenTest(jobDirectory):
    return bool(SubJobXMLList.countSubJobDirs(jobDirectory))
```
at which point the one place it's used may as well just be:
```python
has_children = has_children or SubJobXMLList.countSubJobDirs(path.dirname(fn))
```
and this function is not needed.

rob-c replied:
(aside I hate casting to bools but c'est la vie)

I was planning to use/change this more in the future and to expand/change it. I'm still keen to keep a function which returns a True/False as it's much clearer what's going on.

milliams commented on python/Ganga/Core/GangaRepository/SubJobXMLList.py:
```diff
@@ -517,3 +495,45 @@ def _private_display(self, reg_slice, this_format, default_width, markup):
+            #    subjob_count = subjob_count + 1
+            #i += 1
+
+        return subjob_count
```
For starters, [`os.listdir` is not guaranteed to be ordered](https://docs.python.org/2.7/library/os.html?highlight=listdir#os.listdir).

Secondly, does this really need to only be sequentially numbered directories? If not, this could just be:
```python
@staticmethod
def countSubJobDirs(jobDirectory):
    return len(dir for dir in listdir(jobDirectory) if dir.isdigit())
    
```

rob-c replied:
In that case I'll amend it to sort
Yes because I'm planning to check if any files are missing in another revision.

If this is so broken in the current release shall we bump a working PR to 6.1.20?

milliams replied:
Making changes here is too late for 6.1.20. It's been working ok so far.

Linux will almost always return a sorted list (other systems may not) but the problem come in with the *way* it might sort it. I guess that subjobs 1-12 may be sorted as
```
1
10
11
12
2
3
4
5
6
7
8
9
```

rob-c replied:
Frankly when you can just sort a list it's much simpler/quicker.

If it's not that broken I'll fix it in another revision then. (We don't support anything other than linux right now anyway :P)

rob-c replied:
There is no need to sort this, why is this coming up? I'lll upload an amended version anyway which does some better testing

rob-c commented:
Added some docs this-morning whilst waiting for a few things to complete. This also contains a more updated form of what I'm wanting from the PR.

rob-c commented:
Hmmm, some failing tests but the thread tracer doesn't seem to like something. I'll try and re-run locally in a bit, running the tests locally for me is still fraught with difficulty.

milliams commented:
It's nothing to do with the thread tracer. If an early test fails very badly then the Python session is left in a weird state meaning that the thread tracer doesn't know whether to start. If you look at the [full logs](https://ganga-ci.cern.ch/jenkins/job/PR%20builder/1048/consoleFull) you'll see that you have a syntax error:
```python
self = <Ganga.Core.GangaRepository.GangaRepositoryXML.GangaRepositoryLocal object at 0x3a5a450>

    def shutdown(self):
        """Shutdown the repository. Flushing is done by the Registry
            Raise RepositoryError
            Write an index file for all new objects in memory and master index file of indexes"""
        from Ganga.Utility.logging import getLogger
        logger = getLogger()
        logger.debug("Shutting Down GangaRepositoryLocal: %s" % self.registry.name)
>       for k in k in self._fully_loaded:
E       UnboundLocalError: local variable 'k' referenced before assignment

python/Ganga/Core/GangaRepository/GangaRepositoryXML.py:241: UnboundLocalError
```
If you can't run the tests locally then we need to sort that out as you really should be running them while developing to avoid issues exactly like this.

rob-c commented:
If you can suggest some way to get SetupProject to play nice with this then I'll happily use that. As it stands I'm having issues.
I'm wanting to spend another day working out why the tests won't run properly for me when there is a test box which runs them in an automated fashion. That's the point of this machine in my mind.

milliams commented:
If I log on to `lxplus` in a clean bash shell and do:
```bash
$ pip install --user pytest  # only need run this once
$ cd ganga  # This is the local git checkout of ganga
$ SetupProject Ganga
$ PYTHONPATH=python:$PYTHONPATH ~/.local/bin/py.test python/GangaLHCb/test/GPI/
```
then it successfully runs the LHCb tests. This also works for running the other Ganga tests if you change the path given to pytest.

rob-c commented:
setupproject changes the version of python does it not? also my problem is with pytest pointing to the ganga source not currently being worked on or accessed via SetupProject. I can run tests on another machine, not ideal but works for my needs

milliams commented:
If you make sure you *pre*pend `python` to the python path (or `.` if you're in the `python` directory) then it should always refer to your local version of Ganga. If you have in the past run `pip install -e .` then you may have a reference to that install in your `~/.local/lib/python2.6/site-packages/` which you can delete.

I believe that since `pip` is using Python 2.6, `pytest` will then be installed as Python 2.6. This means that running the code above will use Python 2.6 for running the tests. If you need to use the LHCb version of Python 2.7 to run the tests then the following will install an appropriate version of `py.test`:
```bash
$ SetupProject Ganga
$ wget https://bootstrap.pypa.io/get-pip.py
$ python2.7 get-pip.py --user
$ pip install --user pytest  # only need run this once
```
you can then run the tests using it with:
```bash
$ cd ganga  # This is the local git checkout of ganga
$ SetupProject Ganga
$ PYTHONPATH=python:$PYTHONPATH ~/.local/bin/py.test python/GangaLHCb/test/GPI/
```

rob-c commented:
yes and that worked but doesn't fully reflect my setup now, as I've mentioned this runs elsewhere offline so I'm testing there, I've got the local bin in my path for other tools I use. It's either yet another complexity or something which requires torching my setup and starting again, either way not something I want to spend time on right now

rob-c commented:
OK,

Well this unravelled to be a list of additional papercuts which all needed to be fixed before the main issue went away.

The main changes in this PR are now:

 1. I don't plan to make a test for the original condition in the PR, it's tricky and I'd like to give it some thought before putting it in
 2. I've now moved all knowledge of which objects have been fully loaded into memory into the Repository and am having the registry correctly poll for this. This fixes a chicken and egg issue that arises otherwise when the situation crops up that an object has to be loaded at startup and then made dirty to trigger a new index to be written in the future.
 3. The main index is removed after it's use on startup and only written again on shutdown. Minor performance penalty in starting a 2nd parallel session but this is minor and jobs don't get out of sync on dropped connections etc. #461
 4. The way of checking the number of subjobs is now a little more resilient as well as having the ability to check for all expected XML to be present on load. This fixes a potential bug later on of XML being missing only when working with a job which is trickier to debug/fix.
 5. I've added a setUpClass to the unittest to fix problems with previous tests making the environment dirty before we start running. This should probably be adopted by any tests not making use of this in some other way but that is for another PR.
 6. I've fixed the test for job resubmit in the tutorial to work on faster machines #503
 7. Documentation, I've added a lot of docstrings to the registry/repo layers which makes life a little easier.

As a result the determination of if a job has subjobs associated with it is now a bit more robust and there is an additional check that all of the XML are present before we load a job from disk. (This still allows for us to recover a corrupt/missing XML in favour of it's backup but we won't load a job with a missing subjob, this indicates some severe event occurred and requires some human interaction).

The loading of the XML from disk can probably be simplified even further in a future but some of the checks simply copied from old code are now fixed/removed as they got in the way of understanding this.

Whilst not just due to changes in this PR it's worth noting that:
Checking if an object is now fully loaded into memory is fully working and works in such a way that a Job object is always the same actual object all the way through the registry/repository and only has actions performed on the same object so we can perform checks using ```id(someJob)``` within this code reliably. (This works in front of the proxy as we also cache the proxy for the lifetime of the session too).

rob-c commented:
retest this please

rob-c commented:
Any objections to merging this branch if it passes the tests?

milliams commented on python/Ganga/Core/GangaRepository/GangaRepositoryXML.py:
```diff
@@ -621,8 +672,7 @@ def _safe_flush_xml(self, this_id):
 
-            has_children = (not self.sub_split is None) and\
-                    (self.sub_split in obj._data) and hasattr(obj, self.sub_split) and len(getattr(obj, self.sub_split)) > 0
+            has_children = hasattr(obj, self.sub_split) and getattr(obj, self.sub_split)
```
Could this be:
```python
has_children = getattr(obj, self.sub_split, False)
```

rob-c replied:
Hmm, it probably should be otherwise non splittable objects should throw an exception here... I wonder why this doesn't crop up in testing?

rob-c replied:
Wait, nvm I think I understand the change you're suggesting, I misread that

milliams commented on python/Ganga/Core/GangaRepository/GangaRepositoryXML.py:
```diff
@@ -708,20 +763,32 @@ def flush(self, ids):
                     self._fully_loaded[this_id] = self.objects[this_id]
 
+                subobj_attr = getattr(self.objects[this_id], self.sub_split, None)
+                sub_attr_dirty = getattr(subobj_attr, '_dirty', False)
```
All `GangaObject`s will have the `_dirty` flag so do we need to check for its existence here?

rob-c replied:
Line above can return None

milliams commented on python/Ganga/Core/GangaRepository/GangaRepositoryXML.py:
```diff
@@ -1065,9 +1184,15 @@ def clean(self):
+            _id = next(id_ for id_, o in self._fully_loaded.items() if o is obj)
+            return True
+        except StopIteration:
+            return False
```
Could this be
```python
def isObjectLoaded(self, obj):
    return obj in self._fully_loaded.values()
```

rob-c replied:
I'll give it a try, I started out with that and got into problems which may or may not have gone away now

rob-c replied:
The tests hang at ```Ganga/test/GPI/TutorialTests.py::TutorialTests::test_c_JobManipulation``` there is at least an extreme performance penalty to doing it the simple way or something isn't behaving how we're expecting it to.

rob-c replied:
I gave that test ~30min on an i5 with an SSD that is probably deadlocking in a strange way, I'd prefer to keep the code as it is now because it works.

milliams replied:
This is strange but I guess it's because the `list` `__contains__` method used equality rather than identity and out equality functions are slow and heavy.

rob-c replied:
I've suspected this for a while, performing a check using this often causes significant performance drops. Which reminds me I should probably change the LHCb splitter code in time to use object ids rather than full string comparisons

milliams commented on python/Ganga/Core/GangaRepository/Registry.py:
```diff
@@ -189,9 +218,18 @@ def stop(self):
 
     def join(self, *args, **kwargs):
+        """
+        Not clear if this is called anywhere
```
This is an unhelpful code comment. `join` is implemented since we are subclassing `Thread` and so must make sure that if `join` is called on it, that it works in a polymorphic way and does the correct thing.

rob-c replied:
I didn't write the code I'm just trying to highlight it needs commenting, I'll copy what you've typed there unless you'd prefer to say something else?

milliams replied:
I'm not sure it needs much of a comment. It's a subclass of `threading.Thread` and this is the [`join` method](https://docs.python.org/2.7/library/threading.html?highlight=threading.thread.join#threading.Thread.join). The only potential addition is that is requests the loop to stop running.

milliams commented on python/Ganga/Core/GangaRepository/Registry.py:
```diff
@@ -222,7 +260,15 @@ class Registry(object):
+            dirty_flush_counter (int): This is no longer used
+            update_index_time (int): This is no longer used
+            dirty_max_timeout (int): This is no longer used
+            dirty_min_timeout (int): This is no longer used
```
This is covered by #415

rob-c replied:
So chicken or egg, which to resolve first?

milliams replied:
I'll go ahead and merge my cleanup PRs since they are smaller and have already been fully reviewed.

rob-c replied:
Aside from my comments on #415 OK. Can you poke me when that finishes as we're going to have to wait for the tests to re-run several times before I can get to work on merging the changes into this PR.

milliams commented on python/Ganga/Core/GangaRepository/SubJobXMLList.py:
```diff
@@ -517,3 +561,52 @@ def _private_display(self, reg_slice, this_format, default_width, markup):
         return ds
+
+    @staticmethod
+    def jobHasChildrenTest(jobDirectory, datafileName):
```
Should this be named `jobHasChildren()`?

rob-c replied:
Given it's running the test for the children existing and this triggers the test to run I'd prefer to have this name, I'll update the docs to reflect this

milliams replied:
That's not how function naming works. The name should reflect the API, not the implementation. If the purpose of this function is to answer the question "does this job have children" then the function should be named `jobHasChildren`. But even then, this seems like a strange place to ask this question; this class is always attached to a job object so the user should ask *that* object if it has children.

This function seems to answering the question above while also checking for the presence of the relevant directories (not raising an error if they're not found, simply printing a warning) but if there's no job directory then saying that the job has no children. That's a very confusing function to know how to use correctly.

If the API user wants to know of the job has children then they should ask *that*. if they want to know if there are any missing subjob directories then they should ask *that*. If they want to check for the presence of the job directory then they should check *that*. Combining all three into one function named `jobHasChildrenTest()` does not help with the complexity of the code.

rob-c replied:
I can change the name of the function to ```checkSubJobXML``` and then have a direct call to determine the number of  children in the XML reader.
The whole point of SubJobXMLList class is to care about how the subjobs are stored/accessed/found when on disk. You can't sensibly ask a parent job without starting to reproduce functionality which belongs here. We used to do it elsewhere but then we had no lazy loading for subjobs at all.

milliams replied:
`checkSubJobXML` is a much better name.

milliams commented on python/Ganga/Core/GangaRepository/SubJobXMLList.py:
```diff
@@ -517,3 +561,52 @@ def _private_display(self, reg_slice, this_format, default_width, markup):
+        if subjob_count == len([_folder for _folder in jobDirectoryList if _folder.isdigit()]):
+            return subjob_count
+        else:
+            raise GangaException("Missing subjobs data file in %s" % jobDirectory)
```
Should this be `GangaException`? I forget the rule exactly but I thought that `GangaException` was reserved for things that the user directly caused (like missing arguments to functions etc.) rather than behind-the-scenes things.

rob-c replied:
In this case shall we open a PR to check all uses of this exception type and to document this?

milliams commented on python/Ganga/test/GPI/TutorialTests.py:
```diff
@@ -106,6 +106,7 @@ def test_c_JobManipulation(self):
         # -- JOBMANIPULATION JOBSLICING STOP
 
         # -- JOBMANIPULATION RESUBMIT START
+        jobs(0).kill()
```
This should be before the `# -- JOBMANIPULATION RESUBMIT START` line so that it doesn't show up in that doc section.

milliams commented:
I've merged the small cleanup PRs so you can go ahead and update this PR now.

rob-c commented:
Now the task tests are locking up after this PR... I'll try and investigate

rob-c commented:
retest this please

rob-c commented:
I'm hoping to look to merge this PR today at some stage unless there are any major changes to be made?

I've changed the unittest to a) reflect the fact  the unittest code was moved without being updated and b) to place the gangadir in the ```ws``` in Jenkins to make it easier to pick up  the output of failed tests through the browser if needed.

I'm not seeing any tests fail and this addresses some XML related bugs so I'd be happier once this has been merged.

milliams commented:
Changing the default `gangadir` to be the workspace directory won't help much as the next run of PR Builder will immediately wipe its contents. Also, this is affecting more than just Jenkins, it will make all test runs put the gangadir into the code working directory, even when running them locally.

rob-c commented:
Yes, this is not a bad thing in my opinion. Do we really need to have the tests here restricted to the old testing path?
I know the PR build will wipe the contents but we've got this issue already with the failed test data not persisting and accessing the file from the ws in Jenkins is the easiest way to get the data imo.
(I'm not for adding timestamps or something to the gangadir path but this requires cleaning up after ourselves which will just get messy.)

Can you suggest a better solution for getting access to the data for a failed build?

milliams commented:
No, I just wanted to make sure we were aware of the pros and cons. I can add an entry to my local `.gitignore` to avoid the gangadir polluting things.

milliams commented on python/Ganga/Core/GangaRepository/SubJobXMLList.py:
```diff
@@ -506,3 +556,54 @@ def _private_display(self, reg_slice, this_format, default_width, markup):
+                        if path.isfile(data_file_path):
+                            subjob_count+=1
+                        else:
+                            if path.isfile(data_file_path+'~'):
```
This `else: if:` could be an `elif:`

rob-c replied:
Yes,  development remnant

rob-c commented:
OK, is there anything else holding up this PR or can I push it once the tests pass?

milliams commented on python/Ganga/Core/GangaRepository/SubJobXMLList.py:
```diff
@@ -239,24 +270,11 @@ def __len__(self):
                 return self._stored_len[1]
 
-        from os import listdir, path
+        subjob_count = 0
```
There's no need for this line. it is redefined a few lines later.
